src/protocols/secure_channel/CASESession.cpp

```diff
@@ -884,12 +880,116 @@ CHIP_ERROR CASESession::SendSigma1()
+                                                  kP256_PublicKey_Length,                      // InitiatorEphPubKey,
+                                                  SessionParameters::kEstimatedTLVSize,        // initiatorSessionParams
+                                                  SessionResumptionStorage::kResumptionIdSize, // resumptionId
+                                                  CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES            // initiatorResumeMIC
+    );
+
+    // the PacketBufferHandler should be empty
+    VerifyOrReturnError(msg.IsNull(), CHIP_ERROR_INCORRECT_STATE);
+
+    msg = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+    tlvWriter.Init(std::move(msg));
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorRandom)), input.initiatorRandom));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorSessionId)), input.initiatorSessionId));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kDestinationId)), input.destinationId));
+
+    VerifyOrReturnError(input.pEphPubKey != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorPubKey)), *input.pEphPubKey,
+                                            static_cast<uint32_t>(input.pEphPubKey->Length())));
+
+    VerifyOrReturnError(input.initiatorMrpConfig != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(
+        EncodeSessionParameters(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorMRPParams)), *input.initiatorMrpConfig, tlvWriter));
+
+    if (input.sessionResumptionRequested)
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kResumptionID)), input.resumptionId));
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kResume1MIC)), input.initiatorResumeMICSpan));
+    }
+
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&msg));
+
+    return CHIP_NO_ERROR;
+}
+
 CHIP_ERROR CASESession::HandleSigma1_and_SendSigma2(System::PacketBufferHandle && msg)
 {
     MATTER_TRACE_SCOPE("HandleSigma1_and_SendSigma2", "CASESession");
-    ReturnErrorOnFailure(HandleSigma1(std::move(msg)));
 
-    return CHIP_NO_ERROR;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Parse and Validate Received Sigma1, and decide next step
+    SuccessOrExit(err = HandleSigma1(std::move(msg)));
+
+    switch (mNextStep)
+    {
+    case Step::kSendSigma2: {
+
+        System::PacketBufferHandle msg_R2;
+        EncodeSigma2Inputs encodeSigma2;
+
+        // TODO verify MATTER_LOG_METRIC locations
```

Comment:
Was this done?

Reply:
Yes, I checked with Anush, and he confirmed that from a timing prespective, its fine to keep `MATTER_LOG_METRIC_BEGIN` before the call to `PrepareSigma2`